CONTRIBUTING.md

```diff
@@ -0,0 +1,409 @@
+
+Because the basic formatting is automated by clang-format, we won't
+enumerate the rules here.
+
+#### File conventions
+
+C++ implementation should be named `*.cpp`. Headers should be named `.h`.
+All headers should contain
+
+    #pragma once
+
+All source files should begin with the copyright and license, which
+can be found in the [LICENSE](LICENSE.md) file (or cut and pasted from
+any other other source file).
+
+For NEW source files, please change the copyright year to the present. DO
+NOT edit existing files only to update the copyright year, it just creates
+pointless deltas and offers no increased protection.
+
+
+#### Identifiers
+
+In general, classes and templates should start with upper case and
+capitalize new words: `class CustomerList;` In general, local variables
+should start with lower case. Macros should be `ALL_CAPS`, if used at all.
+
+If your class is extremely similar to, or modeled after, something in the
+standard library, Boost, or something else we interoperate with, it's ok to
+use their naming conventions. For example, very general utility classes and
+templates (the kind of thing you would normally find in std or boost) should
+be lower case with underscores separating words, as they would be if they
+were standards.
+
+    template <class T> shared_ptr;
+    class scoped_mutex;
+
+Names of data should generally be nouns. Functions/methods are trickier: a
+the name of a function that returns a value but has no side effects should
+be a noun, but a procedure that performs an action should be a verb.
+
+#### Class structure
+
+Try to avoid public data members, although there are some classes that serve
+a role similar to a simple C struct -- a very straightforward collection of
+data members. In these, it's fine to make the data members public and have
+clients set and get them directly.
+
+Private member data should be named m_foo (alternately, it's ok to use the
+common practice of member data foo_, but don't mix the conventions within a
+class).
+
+Private member data that needs public accessors should use the convention:
+
+    void foo (const T& newfoo) { m_foo = newfoo; }
+    const T& foo () const { return m_foo; }
+
+Avoid multiple inheritance.
+
+Namespaces: yes, use them!
+
+#### Third-party libraries
+
+Prefer C++11 `std` rather than Boost, where both can do the same task.
+Feel free to use Boost classes you already see in the code base, but don't
+use any Boost you don't see us already using, without first checking with
+the project leader.
+
+#### Comments and Doxygen
+
+Comment philosophy: try to be clear, try to help teach the reader what's
+going on in your code.
+
+Prefer C++ comments (starting line with `//`) rather than C comments (`/* ... */`).
+
+For public APIs we tend to use Doxygen-style comments (start with `///`).
+They look like this:
+
+    /// Explanation of a class.  Note THREE slashes!
+    /// Also, you need at least two lines like this.  If you don't have enough
```

Comment:
This is no longer true -- I don't know if I misunderstood originally, or if Doxygen has gotten smarter over the years. But the blank commented line is not necessary and Doxygen is perfectly happy with a one-line triple-slash comment.